src/query/tsdb/remote/client.go

```diff
@@ -22,120 +22,163 @@ package remote
 	dialOptions = append(dialOptions, additionalDialOpts...)
-
 	cc, err := grpc.Dial("", dialOptions...)
 	if err != nil {
 		return nil, err
 	}
-	client := rpc.NewQueryClient(cc)
 
+	client := rpc.NewQueryClient(cc)
 	return &grpcClient{
-		client:     client,
-		connection: cc,
+		client:        client,
+		connection:    cc,
+		iteratorPools: iteratorPools,
+		workerPool:    workerPool,
 	}, nil
 }
 
 // Fetch reads from remote client storage
-func (c *grpcClient) Fetch(ctx context.Context, query *storage.FetchQuery, options *storage.FetchOptions) (*storage.FetchResult, error) {
+func (c *grpcClient) Fetch(
+	ctx context.Context,
+	query *storage.FetchQuery,
+	options *storage.FetchOptions,
+) (*storage.FetchResult, error) {
+	iters, err := c.fetchRaw(ctx, query, options)
+	if err != nil {
+		return nil, err
+	}
+
+	return storage.SeriesIteratorsToFetchResult(iters, nil, c.workerPool)
+}
+
+func (c *grpcClient) fetchRaw(
+	ctx context.Context,
+	query *storage.FetchQuery,
+	options *storage.FetchOptions,
+) (encoding.SeriesIterators, error) {
+	request, err := EncodeFetchRequest(query)
+	if err != nil {
+		return nil, err
+	}
+
 	// Send the id from the client to the remote server so that provides logging
 	id := logging.ReadContextID(ctx)
-	fetchClient, err := c.client.Fetch(ctx, EncodeFetchMessage(query, id))
+	mdCtx := EncodeMetadata(ctx, id)
+	fetchClient, err := c.client.Fetch(mdCtx, request)
 	if err != nil {
 		return nil, err
 	}
 
 	defer fetchClient.CloseSend()
-
-	tsSeries := make([]*ts.Series, 0)
+	seriesIterators := make([]encoding.SeriesIterator, 0, initResultSize)
 	for {
 		select {
 		// If query is killed during gRPC streaming, close the channel
 		case <-options.KillChan:
 			return nil, errors.ErrQueryInterrupted
 		default:
 		}
+
 		result, err := fetchClient.Recv()
 		if err == io.EOF {
 			break
 		}
+
 		if err != nil {
 			return nil, err
 		}
-		rpcSeries := result.GetSeries()
-		fResult, err := DecodeFetchResult(ctx, rpcSeries)
+
+		iters, err := DecodeCompressedFetchResponse(result, c.iteratorPools)
 		if err != nil {
 			return nil, err
 		}
-		tsSeries = append(tsSeries, fResult...)
-	}
 
-	return &storage.FetchResult{LocalOnly: false, SeriesList: tsSeries}, nil
-}
+		seriesIterators = append(seriesIterators, iters.Iters()...)
+	}
 
-func (c *grpcClient) FetchTags(ctx context.Context, query *storage.FetchQuery, options *storage.FetchOptions) (*storage.SearchResults, error) {
-	return nil, nil
+	return encoding.NewSeriesIterators(
+		seriesIterators,
+		nil,
+	), nil
 }
 
-// Write writes to remote client storage
-func (c *grpcClient) Write(ctx context.Context, query *storage.WriteQuery) error {
-	client := c.client
-
-	writeClient, err := client.Write(ctx)
+func (c *grpcClient) FetchBlocks(
+	ctx context.Context,
+	query *storage.FetchQuery,
+	options *storage.FetchOptions,
+) (block.Result, error) {
+	iters, err := c.fetchRaw(ctx, query, options)
```

Comment:
do the iterators here contain actual block or do we still do the hack later ? If its the later then maybe add a TODO ?

Reply:
These will just be the raw iterators sent from m3; will need to run block conversion logic, but that should go down the path that calls SeriesIteratorsToFetchResult and FetchResultToBlockResult